lib/Sema/TypeCheckConcurrency.h

```diff
@@ -389,6 +391,26 @@ bool diagnoseNonSendableTypes(
   });
 }
 
+/// Diagnose this sendability error with behavior based on the import of
+/// \p nominal . For instance, depending on how \p nominal is imported into
+/// \p fromContext , the diagnostic behavior limitation may be lower or the
+/// compiler might emit a fix-it adding \c \@preconcurrency to the \c import .
+///
+/// \param nominal The declaration whose import we should check, or \c nullptr
+/// to get behavior for a non-nominal type.
+///
+/// \param fromContext The context where the error will be emitted.
+///
+/// \param diagnose Emit a diagnostic indicating a sendability problem,
+/// with the suggested behavior limitation. Returns \c true
+/// if it did not emit any diagnostics.
+///
+/// \returns \c true if any errors were produced, \c false if no diagnostics or
+/// only warnings and notes were produced.
+bool diagnoseAsNonSendableBasedOn(
```

Comment:
This could probably use a better name—suggestions welcome.

Reply:
`diagnoseAsNonSendableFromContext`? It's not that much better.

Reply:
I ended up with `diagnoseSendabilityErrorBasedOn()`, which is at least more vague about the kind of error we're emitting.